client/asset/dcr/dcr.go

```diff
@@ -0,0 +1,1181 @@
+	}
+	if scriptClass != txscript.ScriptHashTy {
+		return nil, fmt.Errorf("unexpected script class %d", scriptClass)
+	}
+	if numReq != 1 {
+		return nil, fmt.Errorf("unexpected number of signatures expected for P2SH script: %d", numReq)
+	}
+	if len(addrs) != 1 {
+		return nil, fmt.Errorf("unexpected number of addresses for P2SH script: %d", len(addrs))
+	}
+	// Compare the contract hash to the P2SH address.
+	contractHash := dcrutil.Hash160(contract)
+	addr := addrs[0]
+	if !bytes.Equal(contractHash, addr.ScriptAddress()) {
+		return nil, fmt.Errorf("contract hash doesn't match script address. %x != %x",
+			contractHash, addr.ScriptAddress())
+	}
+	return &auditInfo{
+		output:     newOutput(dcr.node, txHash, vout, toAtoms(txOut.Value), wire.TxTreeRegular, contract),
+		recipient:  receiver,
+		expiration: time.Unix(int64(stamp), 0).UTC(),
+	}, nil
+}
+
+// FindRedemption should attempt to find the input that spends the specified
+// coin, and return the secret key if it does.
+//
+// NOTE: FindRedemption is necessary to deal with the case of a maker
+// redeeming but not forwarding their redemption information. The DEX does not
+// monitor for this case. While it will result in the counter-party being
+// penalized, the input still needs to be found so the swap can be completed.
+// For typical blockchains, every input of every block starting at the
+// contract block will need to be scanned until the spending input is found.
+// This could potentially be an expensive operation if performed long after
+// the swap is broadcast. Realistically, though, the taker should start
+// looking for the maker's redemption beginning at swapconf confirmations
+// regardless of whether the server sends the 'redemption' message or not.
+func (dcr *ExchangeWallet) FindRedemption(ctx context.Context, coinID dex.Bytes) (dex.Bytes, error) {
+	txHash, vout, err := decodeCoinID(coinID)
+	if err != nil {
+		return nil, err
+	}
+	txid := txHash.String()
+	// Get the wallet transaction.
+	tx, err := dcr.node.GetTransaction(txHash)
+	if err != nil {
+		return nil, fmt.Errorf("error finding transaction %s in wallet: %v", txHash, err)
+	}
+	if tx.Confirmations == 0 {
+		// This is a mempool transaction, GetRawTransactionVerbose works on mempool
+		// transactions, so grab the contract from there and prepare a hash.
+		rawTx, err := dcr.node.GetRawTransactionVerbose(txHash)
+		if err != nil {
+			return nil, fmt.Errorf("error getting contract details from mempool")
+		}
+		if int(vout) > len(rawTx.Vout)-1 {
+			return nil, fmt.Errorf("vout index %d out of range for transaction %s", vout, txHash)
+		}
+		contractHash, err := dexdcr.ExtractContractHash(rawTx.Vout[vout].ScriptPubKey.Hex, chainParams)
+		if err != nil {
+			return nil, err
+		}
+		// Only need to look at other mempool transactions. For each txid, get the
+		// verbose transaction and check each input's previous outpoint.
+		txs, err := dcr.node.GetRawMempool(chainjson.GRMRegular)
```

Comment:
Is this good enough? Could somebody spend a contract output to fund a ticket?

Reply:
Yeah, I think you're right about the need to check stake txns too.